stringdtype/stringdtype/src/casts.c

```diff
@@ -53,11 +53,17 @@ string_to_string(PyArrayMethod_Context *context, char *const data[],
     npy_intp out_stride = strides[1] / context->descriptors[1]->elsize;
 
     while (N--) {
-        out[0] = ssdup(in[0]);
+        ss *s = empty_if_null(in);
+        out[0] = ssdup(s);
         if (out[0] == NULL) {
             gil_error(PyExc_MemoryError, "ssdup failed");
             return -1;
         }
+
+        if (*in == NULL) {
+            free(s);
+        }
```

Comment:
So, if we definitely don't do reference counting (each array that is not a view owns a copy of the string), then I do still think it makes sense to move the length information into the array data itself.

On this case, how about a pattern off:
```
    const ss *s = get_string(in);
```
And then return a statically allocated empty `ss` (or a once allocated global).  That way you don't need the `free()` call.  Freeing is only needed on store, and there you can do it unconditionally.  The only thing is you must not free what `get_string` returns.  I hope the const will protect a bit and it should at least be a nice hard crash that is easy to track down.